488 how to cite

hyi opened the pull request:
Addresses the issue https://github.com/hydroshare/hydroshare/issues/488. Also deployed this branch to beta.hydroshare.org for testing and feedback. Will do more extensive testing and add more validation if needed. @mauriel133 @ironfroggy @pkdash it'd be great if you can do code review.

hydrobuild commented:
Can one of the admins verify this patch?

mauriel133 commented on theme/static/css/hydroshare_core.css:
```diff
@@ -1473,12 +1473,15 @@ body > .container{
+    min-height: 30px;
+    background:#EEE;
+	border: 1px solid #cccccc;
+  	border-radius: 0px;
```
Give this rule 'padding: 4px' to have the text centered in the middle of the container.

hyi replied:
@mauriel133 Good suggestion. Added padding rule and fixed alignment and spacing as well.

mauriel133 commented on theme/static/css/hydroshare_core.css:
```diff
@@ -1473,12 +1473,16 @@ body > .container{
 
 #citation-text a, .rights-url a{
-    word-break: break-word;
+    word-break:normal;
```
The reason I was using 'word-break: break-word' is to keep the page responsive for mobile devices. Without some part of the input will be hidden when shrinking the page to the minimum.

hyi replied:
Got it. Will revert back to break-word style and push it after I rebase with latest develop branch that has core metadata refactoring just merged which triggered some conflicts with my branch I need to resolve.

hyi commented:
I have done some further work to improve code structure and show metadata validation error on resource landing page rather than raising a ValidationError exception, and I am done with what I wanted to do with this issue now and the code is ready for review. @pkdash @mauriel133 please review when you have a chance. Thanks.

pkdash commented on hs_core/models.py:
```diff
@@ -556,6 +563,16 @@ def create(cls, **kwargs):
+            metadata_obj = kwargs['content_object']
+            metadata_type = ContentType.objects.get_for_model(metadata_obj)
+            if kwargs['type'] == 'isHostedBy' and \
+               Relation.objects.filter(type='isCopiedFrom', object_id=metadata_obj.id, content_type=metadata_type).count() > 0:
```
Better to use exists() that returns bool instead of count()

hyi replied:
Good suggestion. Replaced count() with exists() in all relevant places and pushed the code change.

pkdash commented on hs_core/models.py:
```diff
@@ -556,6 +563,16 @@ def create(cls, **kwargs):
+               Relation.objects.filter(type='isCopiedFrom', object_id=metadata_obj.id, content_type=metadata_type).count() > 0:
+                raise ValidationError('Relation type:%s cannot be created since isCopiedFrom relation already exists.' % kwargs['type'])
+            elif kwargs['type'] == 'isCopiedFrom' and \
+                 Relation.objects.filter(type='isHostedBy', object_id=metadata_obj.id, content_type=metadata_type).count() > 0:
```
Same as above

pkdash commented on hs_core/models.py:
```diff
@@ -566,6 +583,18 @@ def update(cls, element_id, **kwargs):
+            rel = Relation.objects.get(id=element_id)
+            if rel.type != kwargs['type']:
+                if kwargs['type'] == 'isHostedBy' and \
+                     Relation.objects.filter(type='isCopiedFrom', object_id=rel.object_id,
```
Same as above

pkdash commented on hs_core/models.py:
```diff
@@ -566,6 +583,18 @@ def update(cls, element_id, **kwargs):
+                                             content_type__pk=rel.content_type.id).count() > 0:
+                    raise ValidationError('Relation type:%s cannot be updated since isCopiedFrom relation already exists.' % rel.type)
+                elif kwargs['type'] == 'isCopiedFrom' and \
+                     Relation.objects.filter(type='isHostedBy', object_id=rel.object_id,
```
Same as above

pkdash commented on hs_core/models.py:
```diff
@@ -1592,23 +1623,27 @@ def get_xml(self, pretty_print=True):
+                term_ns = self.NAMESPACES['dcterms']
+            terms_type = etree.SubElement(dc_rel_rdf_Description, '{%s}%s' % (term_ns, rel.type))
+
+            if rel.value.lower().find(res_url_root) == 0:
```
This means URI that points to hydroshare resource will be encoded differently than from a URI that points to some other sites. The previous code seems to encode any URI the same way.

hyi replied:
@pkdash I might misunderstand the original code as shown below:
```
if rel.value.lower().find('http://') == 0 or rel.value.lower().find('https://') == 0:
            dcterms_type.set('{%s}resource' % self.NAMESPACES['rdf'], rel.value)
else:     
            dcterms_type.text = rel.value
```
I thought it was intended to encode hydroshare resources differently. If I indeed misunderstood, let me know and I will reinstate the previous code as shown above.

pkdash replied:
The old code just checks to see if the value attribute of the relation element is a URI (which can be any valid URI including ones based on hydroshare) and encodes in one specific way (using rdf:resource) otherwise it encodes differently. The new code will encode only hydroshare based URI (using the rdf:resource). I am not sure which one is correct. I am just pointing the difference. @horsburgh can tell us which one is right.

hyi replied:
@horsburgh @pkdash After giving this a bit of more thought, I feel the new code makes more sense. For example, for the newly added "Copied From" relation, the relation value is a URL that points to a host site where the current HydroShare resource is copied from. With the old code logic, this host site URL will be encoded using rdf:resource, and with the new code, this host url will be encoded directly with the url value, and only hydroshare resource will be encoded using rdf:resource. @horsburgh I will wait for your input regarding this, then either keep the new code, or change it back to the old code accordingly.

horsburgh replied:
@hyi and @pkdash - The "Relation" element should really be able to store any type of relationship with another resource.  It may be a resource in HydroShare, and it may be a resource somewhere else on the web.  So - related resources need not be HydorShare resoruces, but need to have a resolvable URI.  Note that we also have the following text in the science metadata document:

If no URI is available for a related resource, a text string that is appropriately specific may be used to describe the related resource (e.g., a bibliographic citation).

mauriel133 commented:
+1 on layout and js.

pkdash commented on hs_core/migrations/0001_initial.py:
```diff
@@ -235,7 +235,9 @@ class Migration(migrations.Migration):
                 ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                 ('object_id', models.PositiveIntegerField()),
-                ('type', models.CharField(max_length=100, choices=[(b'isPartOf', b'Part Of'), (b'isExecutedBy', b'Executed By'), (b'isCreatedBy', b'Created By'), (b'isVersionOf', b'Version Of'), (b'isDataFor', b'Data For'), (b'cites', b'Cites')])),
+                ('type', models.CharField(max_length=100, choices=[(b'isHostedBy', b'Hosted By'), (b'isCopiedFrom', b'Copied From'), (b'isPartOf', b'Part Of'),
```
Were these changes made by Django as part of running makemigrations?

hyi replied:
@pkdash No, they are not. In fact, Django makemigrations does not result in any change in migration files. I added two new field choices in just for consistency. They are not really needed since it works just fine without changing or adding any migration files. Should we add them or not? Either works, but just to be safer and consistent with other field choices, I thought it is better to add them in given that Django makemigrations does not add any migration files in this case (i.e., addition of two new choices for the type field). What do you think?

pkdash replied:
I am assuming manual editing of migration files is probably not recommended. @ironfroggy any thoughts?

ironfroggy replied:
Migrations that have already been shared and run on ohter systems (production, staging, or dev machines) should never be changed. We did that once in a difficult circumstance and need to avoid ever doing it again.

hyi commented:
@ironfroggy @pkdash I reverted back the original migration file and re-run makemigrations, and this time Django did produce a new migration file. Not sure why it did not produce new migration file initially, but the new migration file Django generates looks correct. So I have pushed new migration file and reverted back the original migration file. Please take a look at it and see whether migrations look correct now. Thanks.

hyi commented:
@pkdash @horsburgh OK. Understand now. @pkdash Have reinstated the old code logic for URI encoding and pushed the code change. I believe I have addressed all your comments. Please take a look again and see whether you can give a +1 now.

hyi commented:
@pkdash Just pushed an update - an oversight on my part, should have removed >0 in the first place. Thanks for catching it. See whether everything looks good now.

pkdash commented:
+1
